nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MonitorActivity.java

```diff
@@ -201,17 +253,57 @@ public void process(final OutputStream out) throws IOException {
             }
         } else {
             session.transfer(flowFiles, REL_SUCCESS);
+            updatedLatestSuccessTransfer = now;
             logger.info("Transferred {} FlowFiles to 'success'", new Object[]{flowFiles.size()});
 
-            final long inactivityStartMillis = latestSuccessTransfer.getAndSet(now);
+            final long latestStateReportTimestamp = latestReportedNodeState.get();
+            if (SCOPE_CLUSTER.equals(monitoringScope)
+                    && (now - latestStateReportTimestamp) > (thresholdMillis / 3)) {
+                // We don't want to hit the state manager every onTrigger(), but often enough to detect activeness.
+                try {
+                    final StateManager stateManager = context.getStateManager();
+                    final StateMap state = stateManager.getState(Scope.CLUSTER);
+
+                    final Map<String, String> newValues = new HashMap<>();
+
+                    // Persist attributes so that other nodes can copy it
+                    if (copyAttributes) {
+                        newValues.putAll(flowFiles.get(0).getAttributes());
+                    }
+                    newValues.put(STATE_KEY_LATEST_SUCCESS_TRANSFER, String.valueOf(now));
+
+                    if (state == null || state.getVersion() == -1) {
+                        stateManager.setState(newValues, Scope.CLUSTER);
+                    } else {
+                        // If this returns false due to race condition, it's not a problem since we just need
+                        // the latest active timestamp.
+                        stateManager.replace(state, newValues, Scope.CLUSTER);
```

Comment:
While it usually is the case, we can't assume that the new state has a later timestamp. It could be case that one node is much more taxed than another and it takes longer to transmit state back to the ZooKeeper cluster (where clustered state is stored). This should continually check/attempt to replace until it succeeds or the stored state is later than the current timestamp.